### ISSUE TITLE
Local volumes

### ISSUE THREAD

**Issue by sverhagen** [rhuss/docker-maven-plugin#133]
I would like to be able to do this:

    <volume>${project.build.directory}/myService/logs:/logs</volume>

There seems to be two reasons that this does not work:

1. `DOCKER> cannot bind mount volume: X volume paths must be absolute. (Internal Server Error: 500)`. Where `X` is the drive letter that my Maven project is located at. I'm on Windows, the `${project.build.directory}` variable expands to something like `X:\folder\folder...`, and it appears that the `:` is interpreted as the stanza separator.
2. While `${project.build.directory}` is local to me, my Docker instance isn't. I'm still on Windows, running the Docker daemon in a VM, and *that's* where it's trying to create the folder. In other words, `<volume>/tmp/myService/logs:/logs</volume>` works fine, but it creates it in `/tmp` in the VM, not on my Windows machine, which isn't ideal (in the sense that such build artifacts ideally end up in Maven's `target/` folder).

I'd be curious if we could improve upon this :smile: Thanks!

**Comment by sverhagen** [comment]
Only 13 minutes later I have gained some new insights on this. 1. is not a problem as long as 2. is a problem :smile:

And 2. is a problem because the plugin doesn't really know where my Docker daemon is running, and it would probably need very specific information about the machine that is running the Docker daemon (Boot2docker, VM, etc.) in order to set up a second mount or something from the Docker host to my local machine. I don't see how this is going to work, but perhaps someone else has great insights around this.

**Comment by chris-snider** [comment]
If you are using boot2docker or a vagrant machine, you can mount a path from your Windows system to the VM, then use that reference point as a mount to the running container via the –v argument.

Example, I run a vagrant machine using Virtual Box and mount my development projects folder (C:\projects\currentproject) to the VM as /myproj.  I can then mount this into the container as –v /myproj:/container/path/to/mount.

With this I can have the container write logs etc to the /container/path/to/mount and see/read them from the main Windows interface.

The Boot2docker I have used in the past also uses a Virtual Box VM that I can mount my Windows folders through.

Chris Snider
Senior Software Engineer
Intelligent Software Solutions, Inc.
[Description: Description: Description: cid:image001.png@01CA1F1F.CBC93990]

**Comment by sverhagen** [comment]
Thank you for that, and yes, that's all true. I was hoping for a solution that let my build (folder) be self-contained. Having a mount somewhere below `target/` may not be ideal. Also, it'd put another requirement on the environment of whomever is building this (trying to KISS that as much as possible). So I might just end up with something like `/myService/logs` and "mount it from your docker host to your developer machine if you choose to do so". At this point I might see more value in having a `docker:cp` goal that I could use to copy the logs from the container to somewhere below `target/` (prior to me running `docker:stop`). *Just brainstorming at this point...*

**Comment by rhuss** [comment]
I see your use case, but it's really hard to achieve this without some very specific assumptions. The plugin talks with the Docker host only via the remote REST API. So it cannot be assumed that the Docker host is local (and for your use case it isn't), all we can do is to offer to mount a directory from the Docker host. Building a bridge from the docker host to your local filesystem is IMO out of scope for this plugin. Even a `docker:cp` is not feasible without assuming e.g. an SSH connection (there is nothing like a "copy" from the local system to the Docker host in the Docker remote API).

A probably complete bogus idea would be to use the maven plugin from within a docker container itself. But then plugin still runs remotely (within the container) to the Docker host, so there are probably the same issues. 

To summarise, I want to keep this plugin focussed and as much as possible agnostic from the concrete Docker setup (so no special handling for Vagrant or Boot2Docker, just plain REST calls to a (potentially) remote Docker host). So, if you don't mind, I would like to close the issue. 

Feel free to reopen in case.

**Comment by sverhagen** [comment]
Agreed.
